### ISSUE TITLE
version 1.1.1L crashes with null pointer access in ERR_load_strings()

### ISSUE THREAD

**Issue by yusshi** [openssl/openssl#17971]
Hi all,
version: 1.1.1L
Built with option: no-err
Callstack:
pthread_rwlock_trywrlock(NULL)
pthread_rwlock_timedwrlock(NULL)
CRYPTO_THREAD_write_lock(NULL)
ERR_load_strings()
xmlSecOpenSSLInit()

Original openssl code:
```
int ERR_load_ERR_strings(void)
{
**#ifndef OPENSSL_NO_ERR**
    if (!RUN_ONCE(&err_string_init, do_err_strings_init))
        return 0;

    err_load_strings(ERR_str_libraries);
    err_load_strings(ERR_str_reasons);
    err_patch(ERR_LIB_SYS, ERR_str_functs);
    err_load_strings(ERR_str_functs);
    build_SYS_str_reasons();
#endif
    return 1;
}
```

If changed to:
```
int ERR_load_ERR_strings(void)
{
    if (!RUN_ONCE(&err_string_init, do_err_strings_init))
        return 0;

**#ifndef OPENSSL_NO_ERR**
    err_load_strings(ERR_str_libraries);
    err_load_strings(ERR_str_reasons);
    err_patch(ERR_LIB_SYS, ERR_str_functs);
    err_load_strings(ERR_str_functs);
    build_SYS_str_reasons();
#endif
    return 1;
}
```
Crash does not happen. Is this modification a valid fix?

Thanks,
Yusshi

**Comment by bernd-edlinger** [comment]
Your proposed patch looks reasonable.
I'd suggest to open a pull request.

**Comment by paulidale** [comment]
If you are building without errors, why would you call `ERR_load_ERR_strings()`?

**Comment by mspncp** [comment]
I slightly disagree with @paulidale: wrong usage should not result in a weird crash, so it is indeed a bug imho. But it needs to be fixed elsewhere. According to the call stack,  `ERR_load_ERR_strings()` 

https://github.com/openssl/openssl/blob/739d2bdfba536ff59e8444eb4295b53288ac5caf/crypto/err/err.c#L346-L359

is initially called by `ERR_load_strings()`:

https://github.com/openssl/openssl/blob/739d2bdfba536ff59e8444eb4295b53288ac5caf/crypto/err/err.c#L361-L369

The former is a no-op which returns TRUE, so the latter shouldn't continue calling `err_patch()` and `err_load_strings()`.  This is the true reason why it crashes, because the lock has been initialized and is still NULL (see callstack).

https://github.com/openssl/openssl/blob/739d2bdfba536ff59e8444eb4295b53288ac5caf/crypto/err/err.c#L299


See #17974 and #17975

**Comment by bernd-edlinger** [comment]
Apparently there is also an error in the calling application, because the first parameter "lib" of
`ERR_load_strings()`  should be the result of `ERR_get_next_error_library()`, which in turn
would initialize the `err_string_lock`, so the crash is most likely the result of an invalid parameter passing.

**Comment by mspncp** [comment]
`ERR_get_next_error_library()` should actually be a no-op in the `no-err` case, too. And `err_string_lock` should not be defined at all. See my other https://github.com/openssl/openssl/pull/17975#issuecomment-1080529490.

**Comment by yusshi** [comment]
Thanks for all the responses.
Please correct me if I am wrong. My understanding for no-err is just not to populate the internal error lib/strings of openssl code, but still support external code, such as xmlsec1/curl, etc, to register error lib/strings. So guard all the lines in ERR_load_ERR_strings() may not serve the purpose.

Trigger of this crash is that xmlsec code does not call ERR_get_next_error_library() to get lib number, but use a fixed number instead. 
#define XMLSEC_OPENSSL_ERRORS_LIB                       (ERR_LIB_USER + 57)
We can patch xmlsec to avoid the crash, but openssl should not crash in this case.

**Comment by mspncp** [comment]
> Please correct me if I am wrong. My understanding for no-err is just not to populate the internal error lib/strings of openssl code, but still support external code, such as xmlsec1/curl, etc, to register error lib/strings. So guard all the lines in ERR_load_ERR_strings() may not serve the purpose.

The documentation is not very verbose about this option. The only location I could find is in the INSTALL file:

> **[no-err](https://github.com/openssl/openssl/blob/master/INSTALL.md#no-err)**
>
> Don't compile in any error strings.

Given the fuzziness of the explanation, one could interpret it in the way you understood it.

But if you are right, this would be the most ridiculous OpenSSL compile time option I've encountered so far: the only reasonable justification for the `no-err` option is to save as much precious space as possible in a limited environment. What sense does does it make to apply all the effort to remove all OpenSSL error strings, just to keep third party error strings afterwards?

If people configure OpenSSL without support for error strings, then that's exactly what they should get: no error strings, only error codes, basta.

**Comment by mspncp** [comment]
> Trigger of this crash is that xmlsec code does not call ERR_get_next_error_library() to get lib number, but use a fixed number instead.
> #define XMLSEC_OPENSSL_ERRORS_LIB (ERR_LIB_USER + 57)
> We can patch xmlsec to avoid the crash, but openssl should not crash in this case.

@yusshi would you please be so kind and verify whether  #17975 fixes your crash (without patching xmlsec)?

**Comment by yusshi** [comment]
@mspncp , Sure we can test the patches. Will update when test is done.

**Comment by yusshi** [comment]
@mspncp Tested the patches of yesterday's version, crash does not happen any more. Looks like you are adding more changes.

**Comment by mspncp** [comment]
Sorry @yusshi for asking you prematurely to test. However, now my PR is complete and I believe that the last changes I made don't affect you:

- Instead of removing `err_free_strings_int()` completely, I left the function as a no-op, [as requested](https://github.com/openssl/openssl/pull/17975#discussion_r838330828) by @t8m  (commit [f056557](https://github.com/openssl/openssl/pull/17975/commits/f056557a91be6959ad091bf8e1625c99f4876c8e)).

- I turned `ERR_get_next_error_library()` into a no-op, to get rid of the `err_string_lock` (commit [8a83740](https://github.com/openssl/openssl/pull/17975/commits/8a837406b88b3d6a7a441eda5835cbacde5426ac)).

Feel free to retest, but don't feel obliged to do it for me.

**Comment by yusshi** [comment]
@mspncp Any ETA when #17975 PR can be merged?

**Comment by mspncp** [comment]
It still needs a second approval. But it will certainly be in the next release.  If you need it right now, just fetch the branch.

**Comment by mspncp** [comment]
@yusshi pull request #17975 has just been [merged to the 1.1.1 stable branch](https://github.com/openssl/openssl/pull/17975#issuecomment-1094125145). Pull request #17974 will hopefully follow soon.Thank you for reporting the issue!

**Comment by yusshi** [comment]
Thanks @mspncp for keeping me updated.
